sftp/location.go

```diff
@@ -0,0 +1,166 @@
+package sftp
+
+import (
+	"net/url"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/graymeta/stow"
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+type location struct {
+	// we keep config here so that we can access the server information (username/host)
+	// when constructing the item urls.
+	config     *conf
+	sshClient  *ssh.Client
+	sftpClient *sftp.Client
+}
+
+// CreateContainer creates a new container, in this case a directory on the remote server.
+func (l *location) CreateContainer(containerName string) (stow.Container, error) {
+	if err := l.sftpClient.Mkdir(containerName); err != nil {
+		return nil, err
+	}
+
+	return &container{
+		location: l,
+		name:     containerName,
+	}, nil
+}
+
+// Containers returns a slice of the Container interface, a cursor, and an error.
+func (l *location) Containers(prefix, cursor string, count int) ([]stow.Container, string, error) {
+	infos, err := l.sftpClient.ReadDir(".")
+	if err != nil {
+		return nil, "", err
+	}
+
+	sort.Slice(infos, func(i, j int) bool { return infos[i].Name() < infos[j].Name() })
+
+	var cont []stow.Container
+	for i, v := range infos {
+		if !v.IsDir() {
+			continue
+		}
+
+		if !strings.HasPrefix(v.Name(), prefix) {
+			continue
+		}
+
+		if v.Name() <= cursor {
+			continue
+		}
+
+		cont = append(cont, &container{
+			location: l,
+			name:     v.Name(),
+		})
+
+		if len(cont) == count {
+			// Check if any of the remaining entries are directories. If they are, then return
+			// v.Name() as the cursor. This prevents us from serving up an empty page of
+			// containers if all of the entries after v.Name() are files.
+			for j := i + 1; j < len(infos); j++ {
+				if infos[j].IsDir() {
+					return cont, v.Name(), nil
+				}
+			}
+			return cont, "", nil
+		}
+	}
+
+	return cont, "", nil
+}
+
+// Close closes the underlying sftp/ssh connections.
+func (l *location) Close() error {
+	var errs error
```

Comment:
one thing I wish that multierror provided was way to provide a fmt to the error string... bummer it doesn't allow you to set that format or something of that nature.

would be awesome to replace this code with a type that did multierror and provided a fmt property for printing it in whatever format desired, and have all the proper context. did a scrappy little number to achieve something close here in the control-api in the search stuff:

https://github.com/graymeta/mf2/blob/master/services/control_api/handler.go#L92

Reply:
Yeah, would be a nice to have, especially since this is in the public stow and not our private stuff. I've been dabbling with a custom errors type that would integrate well with the structured logging key/val pairs stuff, but it's still a ways out. We can always revisit in the future.